### PR TITLE
[WIP] Frame Grid

### PR DESCRIPTION
**Pull request by zurbrandon**
Taking the best parts of the Foundation for Apps grid and adding it into sites as an optional extension of the grid.

Meant to work in conjunction with the Foundation for Sites grid or flex grid, but act as the outer shell of a complex web app layout.

This WIP is mainly complete for features of the grid overall and the docs should be about 75% there. Still need some additional browser testing and some improvements to the docs demos. Could also benefit from a template to better show the power of deep nesting with this grid.

Still to do in the wishlist would be the Panel functionality we had developed for Apps. Where on smaller screens a frame-block or frame-group could be made fixed position and triggered as a panel on from off screen. This syntax could use some help and the actual trigger would be need to created using JS. This however was a HUGE help in the Foundation for Apps grid, as it allowed me to have 2-3 columns share horizontal space on large screens, but then have less important content called from off screen on smaller screens.

Looking for help pushing this past the finish.

### CODE REVIEW

**Comment by kball** [comment]
Thanks @zurbrandon for getting this together before you depart!  @zurb/yetinauts lets pick this up and get it to happen for 6.4.  @brettsmason given all of the learning you had on off canvas recently I have a suspicion you might have some thoughts on the panels idea

**Comment by zurbrandon** [comment]
The Functionality is pretty set in apps, but i think the syntax could use some change. @brettsmason This is much easier than the off-canvas since you don't need to transition the body too. Just push the div off on small and translate it on when triggered.

**Comment by paxperscientiam** [comment]
@zurbrandon Really great to see this coming to Foundation for Sites -- great work! Up until now, I've been using custom mixins to handle vertical "blocking". 

I'm playing around with your code and it seems that the `order` classes are not yet in source. That is, they at least do not appear in compiled CSS.

File: scss/grid/_apps-grid-modifiers.scss [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+/// Set the alignment of blocks within a grid.
+///  - left: Items align to the left.
+///  - right: Items align to the right.
+///  - center: Items align to the center.
+///  - justify: Items are spaced equally apart so they occupy the space of the entire grid.
+///  - spaced: Items are given equal space to their left and right.
+/// @param {string} $align - Alignment to use.
+/// @output An appropriate justify-content value.
+@mixin grid-align($align: left) {
```
Comment (DaSchTour) [inline scss/grid/_apps-grid-modifiers.scss]: Similar to flex-align from util/flex

File: scss/grid/_apps-grid-modifiers.scss [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+////
+/// @group frame
+////
+
+/// Define the size of a grid block. Blocks are flex items. By default, they stretch to fill all available space, based on the size of sibling blocks. This is the "expand" behavior.
+/// If set to "shrink", the block will contract and only fill as much space as it needs for its content.
+/// If set to a number, the block will be given a percentage width, based on the total number of columns (12 by default). Percentage widths don't work if a block is inside a vertical grid.
+/// @param {number|string} $size - Sizing behavior of the block. Should be expand, shrink, or a number.
+/// @output The flex-basis, flex-grow, and flex-shrink properties.
+@mixin grid-size($size: expand) {
```
Comment (DaSchTour) [inline scss/grid/_apps-grid-modifiers.scss]: Similar to flex-grid-column from grid/flex-grid

File: scss/grid/_apps-grid-modifiers.scss [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  }
+  @else if ($size == expand) {
+    flex: 1 1 auto;
+  }
+}
+
+/// Set the orientation of blocks within this block. The grid is re-oriented by changing the flex direction of the block.
+/// @param {string} $orientation - Direction of the grid, either horizontal or vertical.
+/// @output A flex-flow property to match the direction given.
+@mixin grid-orient($orientation: horizontal) {
```
Comment (DaSchTour) [inline scss/grid/_apps-grid-modifiers.scss]: Similar to flex-direction from util/flex

File: scss/grid/_apps-grid-modifiers.scss [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    justify: space-between,
+    spaced: space-around,
+  );
+  justify-content: map-get($options, $align);
+}
+
+/// Set the source order of a block. Items with lower numbers appear first. If multiple items have the same number, the one in the HTML first will appear first.
+/// @param {number} $order - Position in source order.
+/// @output An order property.
+@mixin grid-order($order: 0) {
```
Comment (DaSchTour) [inline scss/grid/_apps-grid-modifiers.scss]: The same as flex-order from util/flex

**Review by DaSchTour — CHANGES_REQUESTED** [review]
There are some mixins that could be taken from grid/flex-grid or util/flex

**Comment by brettsmason** [comment]
@zurbrandon sorry for the delay - I'll take a look at this tomorrow night and get back to you.
Looks like @DaSchTour and @paxperscientiam have both been actively looking at it and given great feedback already!

**Comment by kball** [comment]
So you guys know, @zurbrandon is no longer going to be able to put substantial time into Foundation - he hustled to get as much of this in place as possible before he was leaving.

What this means is, if anyone is *excited* about taking this on, improving, polishing, etc... it's wide open! 😁

**Comment by kball** [comment]
@ncoden are you still interested in doing the finishing work here?

File: docs/pages/frame-grid.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+  - scss/grid/apps-grid-modifiers.scss
+tags:
+  - apps grid
+---
+
+### The Grid in Action
+
+The new grid is a complex beast, with a lot of different options, especially when it comes to creating responsive layouts. Check out the sample version below to see how the frame-grid can be used to create complex app layouts.
+
+<div class="frame-grid-sample vertical frame">
```
Comment (corysimmons) [inline docs/pages/frame-grid.md]: Doesn't this need wrapped in triple quotes?

**Comment by kball** [comment]
@brettsmason @corysimmons Did we more or less decide this was being taken care of by the new grid in https://github.com/zurb/foundation-sites/pull/10017 ?

**Comment by corysimmons** [comment]
tbh I'm not exactly sure what Brett decided on the vertical stuff. Is that going into #10017 ?

Either way, this one can probably be closed and Brandon can raise feature requests in #10017

**Comment by kball** [comment]
After digging through this again, I'm pretty convinced that all of the functionality achieved within this will be a part of @brettsmason's new grid, but going to leave this open until that is a little more complete so we can compare

**Comment by kball** [comment]
Closing this as the core ideas were incorporated into the XY grid

**Pull request closed without merging**